trace locals

bavardage opened the pull request:
==COMMIT_MSG==
Introduce trace locals. Each trace gets its own independently settable copy of the variable. 
==COMMIT_MSG==

Commit:
trace locals

carterkozak commented on tracing/src/main/java/com/palantir/tracing/TraceState.java:
```diff
@@ -39,6 +41,8 @@ final class TraceState implements Serializable {
     @Nullable
     private final String forUserAgent;
 
+    private final IdentityHashMap<TraceLocal<?>, Object> traceLocals;
```
We can use a ConcurrentHashMap and declare final equals+hashcode on `TraceLocal` which delegate to super (Object) for identity semantics.

carterkozak commented on tracing/src/main/java/com/palantir/tracing/Tracer.java:
```diff
@@ -64,6 +65,10 @@ private Tracer() {}
     // when our observers are modified.
     private static volatile Consumer<Span> compositeObserver = _span -> {};
 
+    private static final Map<String, TraceLocalObserver<?>> traceLocalObservers = new HashMap<>();
+    // see above comment on compositeObserver
+    private static volatile Consumer<TraceState> compositeTraceLocalObserver = _traceState -> {};
```
tracelocal listeners should be fully encapsulated within the TraceLocal object itself (An overridable method, or custom listener interface would handle this nicely)

bavardage replied:
yeah, fair enough - done

carterkozak commented on tracing/src/main/java/com/palantir/tracing/TraceLocal.java:
```diff
@@ -0,0 +1,38 @@
+        return Tracer.getTraceLocalValue(this, initialValue);
+    }
+
+    public void set(T value) {
```
Possibly worth adding `remove()` to avoid `set(null)` calls.

carterkozak commented on tracing/src/main/java/com/palantir/tracing/TraceLocal.java:
```diff
@@ -0,0 +1,38 @@
+
+public final class TraceLocal<T> {
+
+    private final Supplier<T> initialValue;
```
What are your expectations for the initialValue, how should it be used?
In a world with listeners, is there an expectation that listeners are invoked for a TraceLocal with an initial value which has never had `get` or `set` called within a given trace?

bavardage replied:
yeah, maybe the potential for confusion isn't worth the convenience.

I was imagining: if get/set never called, then listeners are not invoked.

For convenience, it would be nice to have thread locals like `TraceLocal<Counter> counter = new TraceLocal(Counter::new)` which you could easily just use like `counter.get().inc()`...

bavardage replied:
What do you think? Worth the potential for confusion?

bavardage replied:
added a test to demonstrate current behaviour

carterkozak replied:
I think the initialValue component is reasonable, but should be documented to describe the edge cases

bavardage replied:
added javadoc (and will add more when adding observers)

Commit:
keep track of whether detached span is top level or not

Commit:
CR feedback

Commit:
fix to string

Commit:
be extra careful

svc-changelog committed:
Add generated changelog entries

bavardage commented on tracing/src/main/java/com/palantir/tracing/TraceLocal.java:
```diff
@@ -0,0 +1,102 @@
+            log.warn("Five or more TraceLocalObservers registered: {}", SafeArg.of("observers", observers.keySet()));
+        }
+
+        observers.put(name, observer);
```
i didn't do the computing the composite single function thing - thoughts on at what point that becomes worthwhile?

carterkozak replied:
I think we can avoid it entirely by supporting exactly one listener interface when these are created. One could provide a listener which itself maintains a collection of observers that may be registered/unregistered over time for the behavior you've implemented, but this model would keep the minimal case less complex.

In terms of sequencing, I think I'd prefer if we split this feature into two PRs: One which adds the concept of a TraceLocal, and a second which expands them to support listeners.

bavardage replied:
ok, i'll strip the listener stuff out

carterkozak commented on tracing/src/main/java/com/palantir/tracing/TraceLocalObserver.java:
```diff
@@ -0,0 +1,21 @@
+package com.palantir.tracing;
+
+public interface TraceLocalObserver<T> {
+    void onTraceComplete(T value);
```
Should this include other attributes from the `RequestState` e.g. `traceId`, `optional<requestI>`, etc? Such values could also be included in the provided threadlocal value if necessary, but I suspect the traceId at minimum would be helpful for reporting (allows us to tie anything that we report back to the larger operation)

bavardage replied:
yeah makes sense, will add that in the second PR adding observers

Commit:
remove observers, for now

Commit:
javadoc

Commit:
remove unused variable

schlosna commented on tracing/src/main/java/com/palantir/tracing/Tracer.java:
```diff
@@ -374,6 +375,42 @@ static String getRequestId(DetachedSpan detachedSpan) {
+            return;
+        }
+
+        if (value != null) {
+            maybeCurrentTrace.getTraceState().getTraceLocals().put(traceLocal, value);
+        } else {
+            maybeCurrentTrace.getTraceState().getTraceLocals().remove(traceLocal);
+        }
```
Is it worth pulling out traceLocals map to local since it is on both paths to shrink method bytecode size? Also nit on avoiding negation:

```suggestion
        Map<TraceLocal<?>, Object> traceLocals = maybeCurrentTrace.getTraceState().getTraceLocals();
        if (value == null) {
            traceLocals.remove(traceLocal);
        } else {
            traceLocals.put(traceLocal, value);
        }
```

carterkozak replied:
Also `set` shouldn't `remove`, it should only set. Remove should be handled separately, there are subtle performance implications.

bavardage replied:
yeah makes sense

schlosna commented on tracing/src/main/java/com/palantir/tracing/TraceState.java:
```diff
@@ -50,6 +54,7 @@ private TraceState(String traceId, @Nullable String requestId, @Nullable String
         this.traceId = traceId;
         this.requestId = requestId;
         this.forUserAgent = forUserAgent;
+        this.traceLocals = new ConcurrentHashMap<>();
```
How much does this allocation add to perf overhead of tracing? Should we lazily initialize this map only when trace locals are used?

carterkozak replied:
:+1: we should use an atomic-reference-field-updater to thread-safely initialize `traceLocals` to a new map as needed

bavardage replied:
done

carterkozak commented on tracing/src/main/java/com/palantir/tracing/TraceLocal.java:
```diff
@@ -0,0 +1,84 @@
+    /**
+     * Sets the value of this trace local for the current trace.
+     */
+    public void set(@Nonnull T value) {
```
Unlike ThreadLocal, this can race because traces are shared across threads. I wonder if we should provide some way to acknowledge replaced values by returning the previous value or taking some form of collector. Likely unnecessary in most cases, but it would be helpful to retain the capability.

bavardage replied:
what do you mean by taking a collector?

carterkozak replied:
either we return the previous value from `set`, or allow you to provide an update function which may take the previous value e.g. `(TraceLocal<Integer>) traceLocal.update(1, (previousValue, newValue) -> previous == null ? newValue : previousValue + newValue);` I think this sort of API is overkill though.

Edge cases returning the original value from `set` include:
* Do we return `initialValue.get()` if the value is unset?
* Should we provide an API to safely update a value if the current value matches our expectation (avoid update races)?
* Do we need a `setIfAbsent` for the common case, to avoid clobbering values across threads? Otherwise setting, and unsetting when our expectations arent met could allow mutations to be consumed by the temporary update value that we roll back.

bavardage replied:
will return original value from set for now (documenting that this returns null if unset, not initialValue.get()). We can add update/setIfAbsent stuff in future if needed.

bavardage replied:
(as in, those are easy additive changes)

carterkozak commented on tracing/src/main/java/com/palantir/tracing/TraceLocal.java:
```diff
@@ -0,0 +1,84 @@
+     * The supplier is thus normally invoked once per trace, but may be invoked again in case of subsequent
+     * invocations of {@link #remove()} followed by get.
+     */
+    public TraceLocal(Supplier<T> initialValue) {
```
Should we require a supplier in all cases? is `T` nullable?

carterkozak replied:
I think we should make the ctor private, and require a builder for TraceLocal initialization. That way we can optionally allow an initial-value-provider (so get operations don't add data to the internal map unnecessarily) and add the listener interface easily in a follow-up.

bavardage replied:
builder seems overkill for now, but I've made ctor private/added static factory methods

carterkozak replied:
How do you envision eventually creating a new TraceLocal with an initial value supplier and a listener?

carterkozak commented on tracing/src/main/java/com/palantir/tracing/Tracer.java:
```diff
@@ -374,6 +375,42 @@ static String getRequestId(DetachedSpan detachedSpan) {
     }
 
+    @Nullable
+    static <T> T getTraceLocalValue(TraceLocal<T> traceLocal) {
```
unused?

carterkozak commented on tracing/src/main/java/com/palantir/tracing/Tracer.java:
```diff
@@ -374,6 +375,42 @@ static String getRequestId(DetachedSpan detachedSpan) {
+        return (T) maybeCurrentTrace
+                .getTraceState()
+                .getTraceLocals()
+                .computeIfAbsent(traceLocal, _key -> initialValue.get());
```
Let's eagerly transform the `Supplier<T>` into `Function<?, T>` in the TraceLocal, that way we don't need to allocate lambdas for each `get` invocation.

carterkozak commented on changelog/@unreleased/pr-1007.v2.yml:
```diff
@@ -0,0 +1,7 @@
+type: improvement
+improvement:
+  description: Introduce trace locals. Each trace gets its own independently settable
+    copy of the variable. Observers can be registered, which get called with the value
+    of the trace local when a trace completes.
```
```suggestion
    copy of the variable.
```

bavardage committed:
Update tracing/src/main/java/com/palantir/tracing/Tracer.java

Co-authored-by: David Schlosnagle <davids@palantir.com>

bavardage committed:
Update changelog/@unreleased/pr-1007.v2.yml

Co-authored-by: Carter Kozak <ckozak@palantir.com>

Commit:
unused

carterkozak commented on tracing/src/main/java/com/palantir/tracing/TraceState.java:
```diff
@@ -79,12 +84,17 @@ String forUserAgent() {
-                + "forUserAgent='" + forUserAgent
-                + "'}";
+                + "forUserAgent='" + forUserAgent + "', "
+                + "traceLocals=" + traceLocals
```
We should not include the traceLocals map in toString because we have no guarantee that folks won't add tokens. Worth adding a test that traceLocal values aren't included in TraceState.toString
```suggestion
                + "traceLocals=" + traceLocals
```

bavardage replied:
done

carterkozak commented on tracing/src/main/java/com/palantir/tracing/Tracer.java:
```diff
@@ -374,6 +375,43 @@ static String getRequestId(DetachedSpan detachedSpan) {
+        } else {
+            traceLocals.put(traceLocal, value);
+        }
+    }
```
I think we can align the code with intent a bit better by exposing a package-private nullable getter for`TraceState` here, then implemeting the rest of this logic within TraceLocal.

bavardage replied:
it increases the potential for concurrency problems though I think - i.e. without being able to use computeIfAbsent on the ConcurrentHashMap

bavardage replied:
could make tracelocal get synchronized?

carterkozak replied:
My preference would be to avoid splitting the complexity between classes, so `Tracer` provides an internal package-private `TraceState getTraceState()` and `TraceState` provides `@Nullable Map<> getTraceLocals()` and `Map<> getOrCreateTraceLocals()`. Then TraceLocal itself can do any threadsafe logic it likes on the map.

Commit:
CR feedback

private constructor for tracelocal, initial value supplier is optional
lazy init of trace locals in trace state

schlosna commented on tracing/src/main/java/com/palantir/tracing/TraceLocal.java:
```diff
@@ -0,0 +1,96 @@
+     * The supplier is thus normally invoked once per trace, but may be invoked again in case of subsequent
+     * invocations of {@link #remove()} followed by get.
+     */
+    private TraceLocal(Supplier<T> initialValue) {
```
```suggestion
    private TraceLocal(@Nullable Supplier<T> initialValue) {
```

bavardage replied:
yeah, already fixed in https://github.com/palantir/tracing-java/pull/1007/commits/ae573c1e4544f389654c20acbf75a3253edb3e52

Commit:
TraceState toString must not contain trace locals

Commit:
set returns previous value

Commit:
refactoring

Commit:
lint

Commit:
bit of javadoc

Commit:
revert

Commit:
package private

Commit:
elide creation in remove

carterkozak commented on tracing/src/main/java/com/palantir/tracing/TraceLocal.java:
```diff
@@ -0,0 +1,149 @@
+            return null;
+        }
+
+        if (traceState.getTraceLocals() == null) {
+            // no trace locals ever set, short circuit (avoid creating the trace local map)
+            return null;
+        }
+
+        return (T) traceState.getOrCreateTraceLocals().remove(this);
```
```suggestion
        Map<TraceLocal<?>, Object> traceLocals = traceState.getTraceLocals();
        if (traceLocals == null) {
            // no trace locals ever set, short circuit (avoid creating the trace local map)
            return null;
        }

        return (T) traceLocals.remove(this);
```

bavardage replied:
oh sure

carterkozak commented on tracing/src/test/java/com/palantir/tracing/TraceTest.java:
```diff
@@ -44,4 +44,14 @@ public void testToString() {
+        trace.getTraceState().getOrCreateTraceLocals().put(traceLocal, "secret-value");
+
+        assertThat(trace.toString()).doesNotContain("secret");
+    }
```
Nice

carterkozak commented on tracing/src/main/java/com/palantir/tracing/TraceLocal.java:
```diff
@@ -0,0 +1,149 @@
+        } else {
+            // eagerly transform supplier to avoid allocation per invocation
+            // (computeIfAbsent takes a Function)
+            this.initialValue = _ignored -> initialValue.get();
```
Let's add some nullness validation here -- always easier in an initial release than adding new validation later on.
```suggestion
            this.initialValue = _ignored -> Preconditions.checkNotNull(initialValue.get(), "TraceLocal initial value must not be null");
```

carterkozak reviewed (commented):
I'm happy with this pending the last couple of nits. @schlosna Anything from your end?

bavardage committed:
Update tracing/src/main/java/com/palantir/tracing/TraceLocal.java

Co-authored-by: Carter Kozak <ckozak@palantir.com>

bavardage committed:
Update tracing/src/main/java/com/palantir/tracing/TraceLocal.java

Co-authored-by: Carter Kozak <ckozak@palantir.com>

Commit:
fix import

schlosna commented on tracing/src/main/java/com/palantir/tracing/TraceLocal.java:
```diff
@@ -0,0 +1,152 @@
+        } else {
+            // eagerly transform supplier to avoid allocation per invocation
+            // (computeIfAbsent takes a Function)
+            this.initialValue = _ignored ->
+                    Preconditions.checkNotNull(initialValue.get(), "TraceLocal initial value must not be null");
```
Can you confirm my understanding here -- if the supplier/function return null, we'll throw in `TraceLocal::get` not here in constructor? I think we could allow the supplier to return `null` and the value will just not be added to the `traceLocals` map ([`ConcurrentHashMap` allows this](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ConcurrentHashMap.html#computeIfAbsent(K,java.util.function.Function))).

```suggestion
            this.initialValue = _ignored -> initialValue.get();
```

bavardage replied:
yeah, throws in get() currently. Don't have strong feeling about what behaviour is best

carterkozak replied:
My goal with this suggestion is to ensure folks aren't using a supplier which returns null instead of using the factory with no supplier, since that one is more efficient in some cases (`get()` on an empty trace doesn't initialize a new map). I think the behavior is more consistent for users this way as well, but I think we should add a checkNotNull in the `withInitialValue` static factory to throw on ctor when a null supplier is provided.

It's easier to remove the constraint in a future commit than to add a constraint, so my preference was to start as strict as possible.

bavardage replied:
will add the checknotnull in withInitialValue

bavardage replied:
done

carterkozak replied:
I think this will need an approval from you, @schlosna, since some of my suggestions have been merged. I'm going to approve for now given the null check is currently set in the direction of least regret, but I'd like to confirm that you're happy with the current approach -- happy to change/revisit if you have a strong opinion.

schlosna reviewed (commented):
Overall seems good to me. Had one question in TraceLocal

Commit:
ensure initialValue supplier is not null

svc-autorelease commented:
Released [6.16.0](https://github.com/palantir/tracing-java/releases/tag/6.16.0)
